### ISSUE TITLE
Recognition of Collections is inconvenient 

### ISSUE THREAD

**Issue by devmanhinton** [Amicus/end-dash#34]
In order for templates to properly function three main requirements must be satisfied:

1) Template must have a div that is scoped to a name "y" with a child that is scoped to the singular of y

ex. 

``` javascript
     <div class="users-">
        <div class="user-">
          <h1> The User's name is </h1>
          <div class="firstName-"> </div>
        </div>
     </div>
```

2) The collection must be in an anonymous model 

More investigation needed here. In brief this might not come up since we always access collections through other objections (for instance questions from the script model).

To work out of the box, one need do something akin to

``` javascript
    var userCollection = new Backbone.Collection
    var user1 = new User({firstName: "Brian"})
    var user2 = new User({firstName: "Sam"})
    var user3 = new User({firstName: "Justin"})
    userCollection.add([user1, user2, user3])
    var template = EndDash.getTemplate('simpleCollectionBind', {users: userCollection})
```

Note the parameter in the last line where the collection is passed in an object literal. In this case, the name
"users" in both the object literal and template are irrelevant. The same model attribute interpolation can be achieved
using

``` javascript
     <div class="xboxs-">
        <div class="xbox-">
          <h1> The name of the User is </h1>
          <div class="firstName-"> </div>
        </div>
     </div>


    var userCollection = new Backbone.Collection
    var user1 = new User({firstName: "Brian"})
    var user2 = new User({firstName: "Sam"})
    var user3 = new User({firstName: "Justin"})
    userCollection.add([user1, user2, user3])
    var template = EndDash.getTemplate('simpleCollectionBind', {xboxs: userCollection})
```

Relevant End-Dash code:
https://github.com/Amicus/end-dash/blob/master/lib/rules.js#L67
https://github.com/Amicus/end-dash/blob/master/lib/rules.js#L13
https://github.com/Amicus/end-dash/blob/master/lib/rules.js#L62

**Comment by yaymukund** [comment]
Are we planning on doing something like:

``` html
<div data-collection="users">
  <span class="firstName-"></span> <span class="lastName-"></span>
</div>
```

There's no real reason for the `<div class="user-">`, is there?

What's the problem with 2? I don't understand why that's not good.

**Comment by tobowers** [comment]
@yaymukund I've been suggesting data-each or data-each-model

so for the example where you pass in just a collection:

``` html
<ul>
  <li data-each-model class="coolInstanceBro">
    <span class="instanceAttribute-"></span>
  </li>
</ul>
```

and in the case of users being an attribute of the model passed in:

``` html
<ul class="users-">
  <li data-each-model class="coolInstanceBro">
    <span class="instanceAttribute-"></span>
  </li>
</ul>
```

**Comment by yaymukund** [comment]
Why not something like:

``` html
<!-- the parent context is a collection -->
<ul data-each>
  <li><span class="firstName-"></span></li>
</ul>

<!-- `users` is a collection property on the parent context -->
<ul data-each="users-">
  <li><span class="firstName-"></span></li>
</ul>
```

**Comment by tobowers** [comment]
The nice thing about putting it on the instance element is you can now make nice clean semantic html that uses collection properties:

``` html
<div class="users-">
  <p>Total Count: <span class="totalCount-"></span></p>
  <p>Page <span class="currentPage-"></span></p>

  <div data-each>
    <p class="firstName-"><p>
  </div>
</div>
```

**Comment by xcoderzach** [comment]
```
<div class="users-">
  <p>Total Count: <span class="totalCount-"></span></p>
  <p>Page <span class="currentPage-"></span></p>

  <div data-each>
    <p class="firstName-"><p>
  </div>
</div>
```

Would this result in:

```
<div class="users-">
  <p>Total Count: <span class="totalCount-"></span></p>
  <p>Page <span class="currentPage-"></span></p>

  <div data-each>
    <p class="firstName-">ZDawg<p>
    <p class="firstName-">ZSmizzle<p>
    <p class="firstName-">Zach<p>
  </div>
</div>
```

or:

```
<div class="users-">
  <p>Total Count: <span class="totalCount-"></span></p>
  <p>Page <span class="currentPage-"></span></p>

  <div data-each>
    <p class="firstName-">ZDawg<p>
  </div>
  <div data-each>
    <p class="firstName-">ZSmizzle<p>
  </div>
  <div data-each>
    <p class="firstName-">Zach<p>
  </div>
</div>
```

The second is nicer but will probably be a lot more complex wrt sorting, reseting, moving, etc.

**Comment by yaymukund** [comment]
The bad thing about putting it on the instance element is that now you need an extra div whenever you display a child, even if you don't care about collection properties.

Wouldn't nested properties be a better solution?:

``` html
<p>Total: <span class="users.totalCount-"></span></p>
<p>Page <span class="users.currentPage-"></span></p>

<div data-each="users-">
  <p class="firstName-"></p>
</div>
```

**Comment by xcoderzach** [comment]
Nested properties look nice, but they shouldn't be a class property.  Writing that in css is hella weird, b/c it conflicts with the syntax for multiple classes. 

i.e

```
.users.totalCount { // <div class="users totalCount">

}
.users\.totalCount { // <div class="users.totalCount">

}
```

**Comment by tobowers** [comment]
i don't think the nested is as nice, but I'm now thinking that you will always want a wrapper around the collection anyway and doing the data-each with the data-each= might be the cleanest looking?

``` html
<div class="users-">
  <p>Total Count: <span class="totalCount-"></span></p>
  <ul data-each>
    <li><span class="firstName-"></span></li>
  </ul>
</div>

<!-- or -->
  <ul data-each="users-">
    <li><span class="firstName-"></span></li>
  </ul>

```

**Comment by xcoderzach** [comment]
it would be even cooler as:

```
<div class="users-">
  <p>Total Count: #{totalCount}</p>
  <ul data-each>
    <li>#{firstName}</li>
  </ul>
</div>
```

**Comment by devmanhinton** [comment]
^Agreed :+1: +1

**Comment by devmanhinton** [comment]
And this might let variable and attribute reactions be quite similar which would be super nice
And it makes it so that "<property>-", the dash specifically, is only used for changing scope

**Comment by xcoderzach** [comment]
Off in the weeds with implementation details, but... you could do this kinda easily if you remember the index where the # starts, and where the } ends.  And then use Text.splitText() twice to get a reference to the text node, which you can then just replace.  The parser would have to look at text nodes, but that wouldn't be hard.

**Comment by devmanhinton** [comment]
Yeah, I was thinking this would not be too bad actually. Did not get as far implementation as you (ie., Text.splitText() )

This is also sweet cause it would seriously clean up the HTML and make it clear where a template is scoping and interpolating values (similar to what I mentioned above, but a slightly different angle on the benefits. Basically is a boost to code maintainability )

**Comment by xcoderzach** [comment]
We could work on it tonight :wink: :wink: :wink:

**Comment by yaymukund** [comment]
What does `#{property}` have to do with collections?

@tobowers, why don't you think nested properties are nice? I think it's pretty fundamentally broken that we need to add a div to the markup everytime we want to change scope.

I agree with @xcoderzach's concerns about class names and periods. Not sure if there's an easy way around that. Hm.

**Comment by devmanhinton** [comment]
:point_right: :ok_hand: :rocket:  :grin:

**Comment by devmanhinton** [comment]
^ @xcoderzach

**Comment by devmanhinton** [comment]
@yaymukund Because we have been discussing property interpolation with collections
                     And because a major concern is number of divs in endDash and this would significantly reduce the number. 

Random thought but isn't that also nice because then divs are only containers for scopes in EndDash, which fits nicely their role in HTML (namely being containers for DOM elements)?

**Comment by xcoderzach** [comment]
Yeah, discussion about #{prop} should go to #9. 

@yaymukund I agree that adding divs to change scopes is super lame.  There is `data-scope` but that's just a hack and kindof confusing.  

The reason I brought up #{prop} is that it would make #{model.creator.name} possible.  I think that syntax is less shitty.

And, then the concern about class names and periods is mooted, because scoping is only necessary when you have a bunch of variables that share a scope.  If you only have one or two, and there's no element that clearly delineates a scope boundary then you just do 

```
<div class="comment-">
this was posted by: #{comment.poster.first} #{comment.poster.last}
</div>  
```

**Comment by xcoderzach** [comment]
Sorry, bad example

It also makes this possible

```
#{users.length} #{users.page}
```

**Comment by tobowers** [comment]
maybe #{} is used _anytime_ you want to put the value of something somwhere.  This would play nicely into something @devmanhinton brought up earlier this am IRL.

so if you have a container then use -  or in EndDash speak:  to set a scope use a container- but to output a value always do:

``` html
<div class="users-">
  <p>Total Count: #{totalCount}</p>
  <ul data-each>
    <li>#{firstName}</li>
  </ul>
</div>
```

As long as that doesn't fuck up server side rendering that's the nicest thing I've seen here so far.  And does what @yaymukund wants with the dots.

Basically I guess I'm saying _do not even allow_ 

``` html
<div class="user-">
<span class="firstName-"></span>
</div>
```

in favor of 

``` html
<div class="user-">
#{firstName}
</div>
```
